File: app/src/main/java/com/jay/currencyconverter/di/NetworkModule.java
package com.jay.currencyconverter.di;

import com.jay.currencyconverter.model.currencyExchange.bank.Banks;
import com.jay.currencyconverter.model.currencyExchange.nbu.Nbu;
import com.jay.currencyconverter.service.CurrencyExchangeRateApi;

import java.util.List;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Observable;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.jay.currencyconverter.util.Constant.BANKS_EXCHANGE_URL;
import static com.jay.currencyconverter.util.Constant.NBU_URL;

@Module
public class NetworkModule {

    @Provides
    public Observable<Banks> provideBankExchange() {
        return new Retrofit.Builder()
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(BANKS_EXCHANGE_URL)
                .build()
                .create(CurrencyExchangeRateApi.class)
                .getBanksExchangeRate();
    }

    @Provides
    public Observable<List<Nbu>> provideNbuExchange() {
        return new Retrofit.Builder()
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(NBU_URL)
                .build()
                .create(CurrencyExchangeRateApi.class)
                .getNbuExchangeRate();
    }
}

File: app/src/main/java/com/jay/currencyconverter/di/NetworkModule.kt
package com.jay.currencyconverter.di

import com.jay.currencyconverter.annotation.BankExchange
import com.jay.currencyconverter.annotation.NbuExchange
import com.jay.currencyconverter.annotation.NbuExchangeByDateAndCurrency
import com.jay.currencyconverter.api.CurrencyExchangeRateApi
import com.jay.currencyconverter.security.TLSOkHttpClient
import com.jay.currencyconverter.util.Constant.BANKS_EXCHANGE_URL
import com.jay.currencyconverter.util.Constant.NBU_URL
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

@Module
class NetworkModule {

    private val okHttpClient: OkHttpClient? = TLSOkHttpClient.getClient()

    @BankExchange
    @Provides
    fun provideBankExchange(): CurrencyExchangeRateApi {
        return Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .baseUrl(BANKS_EXCHANGE_URL)
            .build()
            .create(CurrencyExchangeRateApi::class.java)
    }

    @NbuExchange
    @Provides
    fun provideNbuExchange(): CurrencyExchangeRateApi {
        return Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .baseUrl("https://bank.gov.ua")
            .apply { if (okHttpClient != null) client(okHttpClient) }
            .build()
            .create(CurrencyExchangeRateApi::class.java)
    }

    @NbuExchangeByDateAndCurrency
    @Provides
    fun provideNbuExchangeByDateAndCurrency(): CurrencyExchangeRateApi {
        return Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
            .baseUrl(NBU_URL)
            .apply { if (okHttpClient != null) client(okHttpClient) }
            .build()
            .create(CurrencyExchangeRateApi::class.java)
    }
}